File: 03-sliding_window/00239-Sliding_Window_Maximum/00239-sliding_window_maximum.py
# 239: Sliding Window Maximum
# https://leetcode.com/problems/sliding-window-maximum/

import collections

class Solution:
    # SOLUTION
    def maxSlidingWindow(self, nums: list[int], k: int) -> list[int]:
        dq = collections.deque()
        result = []
        
        for i, n in enumerate(nums):
            while dq and nums[dq[-1]] < n:
                dq.pop()
            
            dq += i,
            
            if dq[0] == i - k:
                dq.popleft()
            
            if i >= k - 1:
                result += nums[dq[0]],
        
        return result



if __name__ == "__main__":
    o = Solution()

    # INPUT
    nums: list[int] = [1,3,-1,-3,5,3,6,7]
    k: int = 3

    # OUTPUT
    result = o.maxSlidingWindow(nums, k);
    print(result)

File: 03-sliding_window/00239-Sliding_Window_Maximum/00239-sliding_window_maximum.java
// 239: Sliding Window Maximum
// https://leetcode.com/problems/sliding-window-maximum/


import java.util.Deque;
import java.util.LinkedList;


class Solution {
    // SOLUTION
    public int[] maxSlidingWindow(int[] nums, int k) {
        Deque<Integer> dq = new LinkedList<>();
        int[] result = new int[nums.length-k+1];

        int j=0;

        for (int i=0; i<nums.length; i++) {
            if (!dq.isEmpty() && dq.peekFirst() < i-k+1) dq.pollFirst();
            
            while (!dq.isEmpty() && nums[i]>nums[dq.peekLast()]) 
                dq.pollLast();
            
            dq.offer(i);
            
            if (i >= k-1) 
                result[j++] = nums[dq.peekFirst()];
        }

        return result;
    }


    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        int nums[] = {1,3,-1,-3,5,3,6,7};
        int k = 3;

        // OUTPUT
        var result = o.maxSlidingWindow(nums, k);
        System.out.print("["); for (var v : result) System.out.print(v+" "); System.out.println("\b]");
    }
}